### ISSUE TITLE
Dask hanging on long computation when used as joblib backend

### ISSUE THREAD

**Issue by jlopezpena** [dask/dask#2665]
I have an embarrassingly parallel problem that I run under joblib like this:

```python

parallel = Parallel(n_jobs=-1)
results = parallel([delayed(lengthy_function)(x, A, B, C) for x in some_list])
```

the parameters `A, B, C` are the same for all the runs, and they are already pre-computed. A typical run of `lengthy_function` takes about 2 minutes to finish.

When using standard joblin, my code runs just fine. However, when I try to use a dask/distributed backend like this:

```python
with parallel_backend('dask.distributed', scheduler_host='localhost:8786'):
    parallel = Parallel(n_jobs=-1)
    results = parallel([delayed(lengthy_function)(x, A, B, C) for x in some_list])
```

it just hangs without giving me any results. I see some tasks getting dispatched, but nothing ever comes back. In the console where I am running the `dask-worker` instances I just get a warning message (repeated several times):

```
distributed.core - WARNING - Event loop was unresponsive for 1.82s.  This is often caused by long-runn
ing GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
```

Is there anything I can/should do to debug or fix this issue?

**Comment by TomAugspurger** [comment]
Are you able to provide a [small reproducible example](https://stackoverflow.com/help/mcve)?

**Comment by mrocklin** [comment]
I suspect that this is because you are sending large amounts of data up and
down to your cluster.  You might want to try the scatter keyword as
described towards the end of this docpage:
http://distributed.readthedocs.io/en/latest/joblib.html

**Comment by jlopezpena** [comment]
@mrocklin using the scatter keyword has no noticeable effect, dask log shows the "Event loop was unresponsive", and no real CPU use shown. The data I pass to my function is not that large anyway, (dataset is about 3000 rows by 20 columns, rest of arguments are mostly strings or integer flags, the only varying parameter is just a random seed), serializing it for each task should have no noticeable effect (a few extra seconds tops, the body of the function takes several minutes of CPU to run). Weird thing is the extremely low CPU use.

@TomAugspurger I will try to make a minimal example.

**Comment by jlopezpena** [comment]
I have managed to track down the issue a bit further. The job actually finishes with the distributed backend, but takes much much longer than a serial version of the code or multiprocessing backend.

It seems the root of the problem is nested parallelization. I am using the distributed context and scattering my data around the outermost loop, but it seems that there are further calls to parallel contexts below. By manually removing some of them from my own code, I managed to cut down the main running time by about 30% or so, but it is still way slower than the multiprocessing joblib backend. There are still some calls to parallel stuff in deeper parts of the code (scikit-learn code).

I will keep digging.

**Comment by mrocklin** [comment]
We've been chatting with the Joblib devs about handling the nested parallelism case.  It has caused frustration for other applications as well.

When digging in similar situations I have often found the diagnostic dashboard to be of use.  http://distributed.readthedocs.io/en/latest/web.html

**Comment by TomAugspurger** [comment]
@jlopezpena can you try with master joblib / distributed? https://github.com/dask/distributed/pull/1705 may have solved your issue.

**Comment by jlopezpena** [comment]
@TomAugspurger I have recently given a re-trial to joblib/dask with the newer versions and it looks better than before on handling the nested parallellization, but I still manage to lock it somehow. I have managed to reduce the failing case to a simple(ish) function, will try to distill it a bit further and post it here.

For reference, a batch of 100 tasks that takes about 1 min with the loky backend of joblib has been hanging on my machine for two days now. The dask board reports 5 tasks, which is `n_jobs + 1`, so I think re-hitting the same core might be causing a lockdown. The data transfer is minimal (I serialized big datasets to h5 files and just read them from each process when needed), and upon aborting the computation the tracestack tells me that is hanging in this part of the code:

```
~/miniconda3/lib/python3.6/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    749             batch_size = self.batch_size
    750 
--> 751         with self._lock:
    752             tasks = BatchedCalls(itertools.islice(iterator, batch_size),
    753                                  self._backend.get_nested_backend(),
```

**Comment by TomAugspurger** [comment]
@jlopezpena sorry I think I linked to the wrong PR. Can you try out https://github.com/joblib/joblib/pull/914 (not in joblib master yet).

**Comment by jlopezpena** [comment]
Aha, I think I happened to hit two different problems in my code. One of them (the first and most apparent when I opened this issue) was related to nested parallellization, and it got solved by the new joblib dask backend.

The second issue, which is the one I reported in the last comment, I never got to see in the past becase the dask code never made it far enough down the line to hit it, and it seems to happen when there are more tasks than cores. From what I read in the PR you link (and the other threads mentioned there) that seems to be the issue other people are observing, so I think the fix should work for me.

Will try installing a custom joblib with PR914 merged and report.
